supportedBackends/aws/src/main/resources/sidecar/sidecar

```diff
@@ -0,0 +1,117 @@
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+#  HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+#  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+#  IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+
+# We expect the following environment variables:
+#
+# AWS_CROMWELL_TASK_DOCKER_ID
+# AWS_CROMWELL_RC_FILE
+# AWS_CROMWELL_PATH
+# AWS_CROMWELL_LOCAL_DISK
+# AWS_CROMWELL_CALL_ROOT
+
+waitfor() {
+  echo Watching "$1". Will return when "$2" is found
+  inotifywait -r -e close_write,moved_to,moved_from,delete -m "$1" | while read -r path action file; do
+    if [ "$file" = "$2" ]; then
+      echo Target file found - killing inotifywait and returning
+      killall inotifywait # inotifywait will simply hang out with a return like this
+      return
+    fi
+    relativePath=${path#"${localCallRoot}/"}
+    localDiskRelativePath=${path#"${AWS_CROMWELL_LOCAL_DISK}/"}
+    if [ "$action" = "MOVED_TO" ] || [ "$action" = "CLOSE_WRITE,CLOSE" ]; then
+      case $path in
+        ${AWS_CROMWELL_LOCAL_DISK}/${exclude}*)
+          echo skipping change in excluded directory
+          ;;
+        *)
+          aws s3 cp "${path}${file}" "$AWS_CROMWELL_CALL_ROOT/${localDiskRelativePath}${file}"
+      esac
+    fi
+    if [ "$action" = "MOVED_FROM" ]; then
+      case $path in
+        ${AWS_CROMWELL_LOCAL_DISK}/${exclude}*)
+          echo skipping change in excluded directory
+          ;;
+        *)
+          aws s3 rm "$AWS_CROMWELL_CALL_ROOT/${localDiskRelativePath}${file}"
+      esac
+    fi
+  done
+}
+
+###########################################################################
+# The general plan is to do a sync of all workflow files into the container
+# We share the same volumes/mount points of the original, so we can go to
+# town here. Generally, we assume all *inputs* come from
+# AWS_CROMWELL_WORKFLOW_ROOT
+# All *outputs* will eventually end up in AWS_CROMWELL_CALL_ROOT.
+# However, within the container, this root is effectively AWS_CROMWELL_LOCAL_DISK
+#
+# The trick here is that the inputs are also landed inside
+# AWS_CROMWELL_LOCAL_DISK. So we need to exclude the inputs from being
+# synced (both real time and in batch at the end), so we do some fancy
+# stuff for that
+###########################################################################
+# call root example: s3://cromwell-aws/cromwell-execution/...
+# localRoot example: /cromwell_root/cromwell-aws/cromwell-execution...
+# AWS_CROMWELL_PATH example: ...
+localCallRoot="$AWS_CROMWELL_LOCAL_DISK"/"${AWS_CROMWELL_CALL_ROOT#"s3://"}"
+localWorkflowRoot="$AWS_CROMWELL_LOCAL_DISK"/"${AWS_CROMWELL_WORKFLOW_ROOT#"s3://"}"
+suffix="/*"
+exclude=${AWS_CROMWELL_CALL_ROOT#"s3://"}
+exclude=${exclude%%$suffix}
+echo "syncing files from s3"
+echo "source ${AWS_CROMWELL_WORKFLOW_ROOT}, dest $localWorkflowRoot"
+aws s3 sync "$AWS_CROMWELL_WORKFLOW_ROOT" "$localWorkflowRoot"
```

Comment:
> Is this also localizing inputs to this call that are the outputs from upstream calls?

Hmm if that's true the performance and storage costs of localization for large workflows are going to be very heavy IMO

Also, is this assuming that all inputs to the call will be in the s3 workflow root ?
Because even though a lot of time this will be true, it won't always be, specifically the inputs to the workflow itself will be on the user's bucket or even some public bucket, wherever the input data is.

Reply:
Is there a centaur test that shows this outside the s3 workflow root scenario @Horneth? I was assuming I'd need to pass some external input requests as metadata into the container but I didn't see any examples.

Reply:
I found [this one](https://github.com/broadinstitute/cromwell/blob/develop/centaur/src/main/resources/standardTestCases/missing_sub_inputs/missing_sub_inputs.inputs) for GCS/PAPI for example.
But more generally workflow inputs *will be* outside the workflow root for any real workflow, and a job should be able to take inputs from any bucket (modulo permissions of course).
When you say "pass some external input requests as metadata", can you elaborate ? Could we do it for all job inputs instead of just the ones that are outside the workflow root, and avoid having to localize the whole thing ?

Reply:
@elerch  the HaplotypeCaller test also has external inputs -- Khalid posted the inputs file with those examples in the aws cromwell slack.